File: Java/프로그래머스/2/49994. 방문 길이/방문 길이.java


class Solution {
    public int solution(String dirs) {
        int answer = 0;
        int x = 5, y = 5;
        boolean[][] hVisited = new boolean[11][10];
        boolean[][] vVisited = new boolean[10][11];
        
        for (int i = 0; i < dirs.length(); i++) {
            char c = dirs.charAt(i);
            int[] dxy = calDxy(c);
            int nx = x + dxy[0];
            int ny = y + dxy[1];
            
            // 범위를 벗어나면 무시
            if (nx < 0 || nx > 10 || ny < 0 || ny > 10) continue;
            
            if (c == 'L' || c == 'R') {
                int minX = Math.min(x, nx);
                if (!hVisited[y][minX]) {
                    hVisited[y][minX] = true;
                    answer++;
                }
            } else {
                int minY = Math.min(y, ny);
                if (!vVisited[minY][x]) {
                    vVisited[minY][x] = true;
                    answer++;
                }
            }
            
            x = nx;
            y = ny;
        }
        
        return answer;
    }
    
    private static int[] calDxy(char command) {
        if (command == 'U') return new int[]{0, 1};
        if (command == 'D') return new int[]{0, -1};
        if (command == 'L') return new int[]{-1, 0};
        if (command == 'R') return new int[]{1, 0};
        return new int[]{0, 0};
    }
}

File: Swift/프로그래머스/2/49994. 방문 길이/방문 길이.swift
import Foundation

func solution(_ dirs:String) -> Int {
    var answer = 0
    var x = 5
    var y = 5
    
    var hVisited: [[Bool]] = .init(repeating: .init(repeating: false, count: 10), count: 11)
    var vVisited: [[Bool]] = .init(repeating: .init(repeating: false, count: 11), count: 10)
    
    for c in dirs {
        let cToS = String(c)
        let dxy = calDxy(cToS)
        let nx = x + dxy.0
        let ny = y + dxy.1
        
        if nx < 0 || nx > 10 || ny < 0 || ny > 10 {
            continue
        }
        
        if cToS == "L" || cToS == "R" {
            let minX = min(x, nx)
            if !hVisited[y][minX] {
                hVisited[y][minX] = true
                answer += 1
            }
        } else {
            let minY = min(y, ny)
            if !vVisited[minY][x] {
                vVisited[minY][x] = true
                answer += 1
            }
        }
        
        x = nx
        y = ny
    }
    
    return answer
}

func calDxy(_ command: String) -> (Int, Int) {
    switch command {
        case "U":
            return (0, 1)
        case "D":
            return (0, -1)
        case "L":
            return (-1, 0)
        case "R":
            return (1, 0)
        default:
            return (0, 0)
    }
}